components/board/genericlinux/board.go

```diff
@@ -481,27 +474,25 @@ func (b *Board) StreamTicks(ctx context.Context, interrupts []board.DigitalInter
 		i.AddChannel(ch)
 	}
 
-	b.activeBackgroundWorkers.Add(1)
-	utils.ManagedGo(func() {
+	b.workers.AddWorkers(func(cancelCtx context.Context) {
 		// Wait until it's time to shut down then remove callbacks.
 		select {
 		case <-ctx.Done():
-		case <-b.cancelCtx.Done():
+		case <-cancelCtx.Done():
 		}
 		for _, i := range rawInterrupts {
 			i.RemoveChannel(ch)
 		}
-	}, b.activeBackgroundWorkers.Done)
+	})
 
 	return nil
 }
 
 // Close attempts to cleanly close each part of the board.
 func (b *Board) Close(ctx context.Context) error {
 	b.mu.Lock()
-	b.cancelFunc()
-	b.mu.Unlock()
-	b.activeBackgroundWorkers.Wait()
+	defer b.mu.Unlock()
```

Comment:
Yeah, the old version had a subtle bug. The mutex is supposed to protect `b.gpios`, `b.interrupts`, etc. and has nothing to do with `b.workers`. That way, if someone implicitly converts a GPIO pin to a digital interrupt while we're trying to close the board, we don't leak resources. In the old version of the code, it definitely doesn't do that! :flushed: 

I wish we could adopt the following policies:
- if you need to lock a mutex, do it only at the top of a function, never in the middle
- if you lock a mutex, immediately defer unlocking it, rather than unlocking it in a non-deferred way later.
I suspect those two changes would remove entire categories of bugs, at the cost of also removing certain (IMO brittle, bug-prone) design options.